cpp/src/arrow/error_or.h

```diff
@@ -0,0 +1,427 @@
+  /// &&`.
+  ///
+  /// Sets `other` to contain a non-OK status with a`StatusError::Invalid`
+  /// error code.
+  ///
+  /// \param other The ErrorOr object to move from and set to a non-OK status.
+  template <typename U,
+            typename E = typename std::enable_if<std::is_constructible<T, U>::value &&
+                                                 std::is_convertible<U, T>::value>::type>
+  ErrorOr(ErrorOr<U>&& other) : has_value_(other.has_value_) {
+    if (has_value_) {
+      new (&variant_) variant(std::move(other.variant_.value_));
+      other.OverwriteValueWithStatus(
+          Status::Invalid(ErrorOrConstants::kValueMoveConstructorMsg));
+    } else {
+      new (&variant_) variant(std::move(other.variant_.status_));
+#ifndef NDEBUG
+      // The other.variant_.status_ gets moved and invalidated with a Status-
+      // specific error message above. To aid debugging, set the status to a
+      // ErrorOr-specific error message.
+      other.variant_.status_ =
+          Status::Invalid(ErrorOrConstants::kStatusMoveConstructorMsg);
+#endif
+    }
+  }
+
+  /// Move-assignment operator.
+  ///
+  /// Sets `other` to contain a non-OK status.
+  ///
+  /// \param other The ErrorOr object to assign from and set to a non-OK
+  /// status.
+  ErrorOr& operator=(ErrorOr&& other) {
+    // Check for self-assignment.
+    if (this == &other) {
+      return *this;
+    }
+
+    // Construct the variant object using the variant object of the donor.
+    if (other.has_value_) {
+      AssignValue(std::move(other.variant_.value_));
+      other.OverwriteValueWithStatus(
+          Status::Invalid(ErrorOrConstants::kValueMoveAssignmentMsg));
+    } else {
+      AssignStatus(std::move(other.variant_.status_));
+#ifndef NDEBUG
+      // The other.variant_.status_ gets moved and invalidated with a Status-
+      // specific error message above. To aid debugging, set the status to a
+      // ErrorOr-specific error message.
+      other.variant_.status_ =
+          Status::Invalid(ErrorOrConstants::kStatusMoveAssignmentMsg);
+#endif
+    }
+
+    return *this;
+  }
+
+  /// Indicates whether the object contains a `T` value.
+  ///
+  /// \return True if this ErrorOr object's status is OK (i.e. a call to ok()
+  /// returns true). If this function returns true, then it is safe to access
+  /// the wrapped element through a call to ValueOrDie().
+  bool ok() const { return has_value_; }
+
+  /// Gets the stored status object, or an OK status if a `T` value is stored.
+  ///
+  /// \return The stored non-OK status object, or an OK status if this object
+  ///         has a value.
+  Status status() const { return ok() ? Status::OK() : variant_.status_; }
+
+  /// Gets the stored `T` value.
+  ///
+  /// This method should only be called if this ErrorOr object's status is OK
+  /// (i.e. a call to ok() returns true), otherwise this call will abort.
+  ///
+  /// \return The stored `T` value.
+  const T& ValueOrDie() const& {
+    if (!ok()) {
```

Comment:
I think it makes sense to use `ARROW_PREDICT_FALSE`, and everywhere you see `!ok()`.

Reply:
done.